README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Mickaël Canouil’s Analysis Scripts Templates

<!-- badges: start -->

[![License](https://img.shields.io/github/license/mcanouil/analysis-scripts-templates)](LICENSE)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)]()
<!-- [![GitHub
tag](https://img.shields.io/github/tag/mcanouil/analysis-scripts-templates.svg?label=latest%20tag&include_prereleases)](https://github.com/mcanouil/analysis-scripts-templates) -->
<!-- badges: end -->

Script templates to work with omics datasets.

## R

  - [default.R](R/default.R)

### R/functions

  - [tar-crossmap.R](R/functions/tar-crossmap.R)
  - [tar-ewas.R](R/functions/tar-ewas.R)
  - [tar-gwas.R](R/functions/tar-gwas.R)
  - [tar-manhattan.R](R/functions/tar-manhattan.R)
  - [tar-plink2.R](R/functions/tar-plink2.R)
  - [tar-pval\_trans.R](R/functions/tar-pval_trans.R)
  - [tar-qc\_idats.R](R/functions/tar-qc_idats.R)
  - [tar-qc\_plink.R](R/functions/tar-qc_plink.R)
  - [tar-sub\_chunk.R](R/functions/tar-sub_chunk.R)
  - [tar-vep.R](R/functions/tar-vep.R)

### R/scripts

  - [01-design.R](R/scripts/01-design.R)
  - [02-qc\_idats\_snps.Rmd](R/scripts/02-qc_idats_snps.Rmd)
  - [02-qc\_idats.Rmd](R/scripts/02-qc_idats.Rmd)
  - [03-qc\_plink.Rmd](R/scripts/03-qc_plink.Rmd)
  - [04-qc\_impute.Rmd](R/scripts/04-qc_impute.Rmd)
  - [05-ethnicity.R](R/scripts/05-ethnicity.R)
  - [06-omni\_to\_hg38.R](R/scripts/06-omni_to_hg38.R)
  - [07-save\_outputs.R](R/scripts/07-save_outputs.R)
  - [08-vep\_vcf\_docker.R](R/scripts/08-vep_vcf_docker.R)
  - [09-gwas.R](R/scripts/09-gwas.R)
  - [10-ewas.R](R/scripts/10-ewas.R)
  - [11-twas.R](R/scripts/11-twas.R)
  - [12-ora.R](R/scripts/12-ora.R)
  - [13-gsea.R](R/scripts/13-gsea.R)
  - [14-mqtl.R](R/scripts/14-mqtl.R)
  - [15-eqtl.R](R/scripts/15-eqtl.R)
  - [16-eqtm\_estimated\_time.R](R/scripts/16-eqtm_estimated_time.R)
  - [16-eqtm.R](R/scripts/16-eqtm.R)

### R/targets

  - [tar\_crossmap.R](R/targets/tar_crossmap.R)
  - [tar\_ewas.R](R/targets/tar_ewas.R)
  - [tar\_genotype.R](R/targets/tar_genotype.R)
  - [tar\_gsea.R](R/targets/tar_gsea.R)
  - [tar\_gwas.R](R/targets/tar_gwas.R)
  - [tar\_methylation.R](R/targets/tar_methylation.R)
  - [tar\_ora.R](R/targets/tar_ora.R)
  - [tar\_vep.R](R/targets/tar_vep.R)

## Rmarkdown

Note: `assets` directory from
[github.com/umr1283/xaringan-template](https://github.com/umr1283/xaringan-template)
is required\!

  - [ewas\_report.Rmd](Rmarkdown/ewas_report.Rmd)
  - [genotype\_array\_qc\_report.Rmd](Rmarkdown/genotype_array_qc_report.Rmd)
  - [gwas\_report.Rmd](Rmarkdown/gwas_report.Rmd)
  - [methylation\_array\_qc\_report.Rmd](Rmarkdown/methylation_array_qc_report.Rmd)


NEWS.md
# analysis_templates (development version)

* `R/targets/`
  + `tar_crossmap.R`, `targets` setup to upgrade or downgrade genome assembly of VCF files.
  + `tar_ewas.R`, `targets` setup to perform multiple EWAS.
  + `tar_genotype.R`, `targets` setup to compute PLINK files quality-control.
  + `tar_gsea.R`, `targets` setup to compute gene-set enrichment analysis.
  + `tar_gwas.R`, `targets` setup to perform multiple GWAS using PLINK2 software.
  + `tar_methylation.R`, `targets` setup to compute idats (methylation) files quality-control.
  + `tar_ora.R`, `targets` setup to compute over-representation analysis.
  + `tar_vep.R`, `targets` setup to compute VEP annotations.

* `R/functions/`
  + `tar-crossmap.R`, functions, used in `tar_crossmap.R`, to upgrade or downgrade genome assembly of VCF files.
  + `tar-ewas.R`, function, used in `tar_ewas.R`, to perform multiple EWAS.
  + `tar-gwas.R`, function, used in `tar_gwas.R`, to perform multiple GWAS using PLINK2 software.
  + `tar-manhattan.R`, functions to draw Manhattan plot using `ggplot2`.
  + `tar-plink2.R`n function to download PLINK2 binary.
  + `tar-pval_trans.R`, p-value `ggplot2` axis transformation.
  + `tar-qc_idats.R`, functions, used in `tar_methylation.R`, to compute idats (methylation) files quality-control.
  + `tar-qc_plink.R`, functions, used in `tar_genotype.R`, to compute PLINK files quality-control.
  + `tar-sub_chunk.R`, function to create sub-chunk inside a chunk in RMarkdown.
  + `tar-vep.R`, functions, used in `tar_vep.R`, to prepare Docker command and to format VEP output.

* `R/rmarkdown/` (`assets` directory from [github.com/umr1283/xaringan-template](https://github.com/umr1283/xaringan-template) is required)
  + `ewas_report.Rmd`, `xaringan` report template associated with `R/targets/tar_ewas.R` and `R/functions/tar-ewas.R`.
  + `genotype_array_qc_report.Rmd`, `xaringan` report template associated with `R/targets/tar_genotype.R` and `R/functions/tar-qc_plink.R`.
  + `gwas_report.Rmd`, `xaringan` report template associated with `R/targets/tar_gwas.R` and `R/functions/tar-gwas.R`.
  + `methylation_array_qc_report.Rmd`, `xaringan` report template associated with `R/targets/tar_methylation.R` and `R/functions/tar-qc_idats.R`.

* `R/scripts`
  + `**-****.R`,
    - fix project name to not include file extension.
    - trim trailing spaces.
  + `default.R`,
    - default R script.
  + `01-design.R`,
    - skeleton script make batch design for omics (currently "empty").
  + `02-qc_idats.Rmd`,
    - YAML header for Rmarkdown template in `umr1283/dmapaq`.
  + `02-qc_idats_snps.Rmd`,
    - YAML header for Rmarkdown template in `umr1283/dmapaq` (mostly for mQTL analysis).
  + `03-qc_plink.Rmd`,
    - skeleton script or YAML header for Rmarkdown template in `umr1283/dgapaq`.
  + `04-qc_impute.Rmd`,
    - skeleton script or YAML header for Rmarkdown template in `umr1283/dgapaq`.
  + `05-ethnicity.R`,
    - skeleton script to perform ethnicity inference using 1,000 Genomes and HRC imputed VCF files.
  + `06-omni_to_hg38.R`,
    - skeleton R script to upgrade genotyping arrays to GRCh38 genome assembly.
  + `07-save_outputs.R`,
    - skeleton R script to store "sensitive"/"fragile" data to a more secure disk.
  + `08-vep_vcf_docker.R`,
    - skeleton R script with iterative steps to retrieve RSID and gene's symbol using VEP (Docker).
    - add function to prepare docker command and to format output.
  + `09-gwas.R`,
    - skeleton R script to perform a GWAS using `PLINK2`  with multiple traits.
  + `10-ewas.R`,
    - skeleton R script to perform a EWAS using `limma` with multiple traits.
    - tweak & refactor some code.
    - use as.matrix from `data.table`.
  + `11-twas.R`,
    - skeleton R script to perform a GWAS using `DESeq2` with multiple traits.
    - fix #2 (@Ning-L).
    - ensure Ensembl ID does not contain version.
  + `12-ora.R`,
    - add gene symbols for KEGG.
  + `13-gsea.R` and `R/tar_gsea.R`,
    - add genes set and peripheral sets (genes set - core set).
    - add gene symbols in addition to Ensembl, Uniprot or Entrez gene IDs.
    - fix ties using pvalue as secondary parameter in sorting and removing `.Machine$double.eps` for the second value.
  + `14-mqtl.R`,
    - skeleton R script to perform mQTL analysis (methylation Vs. genotypes).
    - fix missing variables.
    - tweak & refactor some code.
    - use as.matrix from `data.table`.
  + `15-eqtl.R`,
    - keleton R script to perform eQTL analysis (expression Vs. genotypes).
    - fix missing variables.
  + `16-eqtm.R`,
    - new expression-methylation analysis.
    - tweak & refactor some code.
    - isoform/transcript or gene level.
  + `16-eqtm_estimated_time.R`, new script to estimate computation time from `16-eqtm.R`'s log.


R/15-eqtl.R
message(timestamp(quiet = TRUE))
### Project Setup ==================================================================================
library(here)
project_name <- sub("(.*)_*\\..*", "\\1", list.files(here(), pattern = ".Rproj$"))
output_directory <- here("outputs", "15-eqtl")
dir.create(output_directory, recursive = TRUE, showWarnings = FALSE, mode = "0775")

vep_directory <- here("outputs", "08-vep_vcf_docker")

run_directory <- "/disks/RUN/Run_XXX/Output/RSEM"

phenotype <- here("docs", "phenotype.xlsx")
ethnicity <- file.path(data_directory, "Omni2.5", paste0(project_name, "_ethnicity.csv"))
exclusion <- file.path(data_directory, "Omni2.5", paste0(project_name, "_QC_exclusion.xlsx"))

seed <- as.numeric(Sys.Date())
cis_window <- 500


### Load Packages ==================================================================================
suppressPackageStartupMessages({
  library(parallel)
  library(tximport)
  library(DESeq2)
  library(matrixStats)
  library(glue)
  library(biomaRt)
  library(httr)
  library(data.table)
})


### Setup biomaRt ==================================================================================
set_config(config(ssl_verifypeer = FALSE)) # Fix SSL check error

species <- "hsapiens_gene_ensembl"
build <- 38
version <- 103

get_mart <- quote(useEnsembl(
  biomart = "ensembl", 
  dataset = species, 
  version = version, 
  GRCh = if (build == 37) build else NULL
))

mart <- try(eval(get_mart), silent = TRUE)
if (inherits(mart, "try-error")) mart <- eval(get_mart)
ensembl_version <- sprintf("GRCh%d-%d", build, version)


### Analysis =======================================================================================
for (rna_level in c("genes", "isoforms")) { 
  local({
    rna_level_name <- unname(c("genes" = "gene", "isoforms" = "transcript")[rna_level])
    
    output_directory <- here("outputs", "14-eqtl", rna_level)
    dir.create(output_directory, recursive = TRUE, showWarnings = FALSE, mode = "0775")
    
    invisible(sapply(c("rnaseq", "covariates", "fastqtl", "fastqtl_annotated", "fastqtl_combined"), function(x) {
      assign(x = paste0("output_", x), value = file.path(output_directory, x), envir = .GlobalEnv)
      dir.create(file.path(output_directory, x), recursive = TRUE, showWarnings = FALSE, mode = "0775")
    }))
    
    
    ### Get QCed Phenotypes ============================================================================
    sample_sheet_qc <- merge(
      x = setDT(read_excel(phenotype)),
      y = fread(ethnicity),
      by = "iid"
    )[j = `:=`(
      "Sample_ID" = iid,
      "#IID" = iid,
      "vcf_id" = iid,
      "group" = as.integer(groupe) + 1L,
      "sex" = c("M" = 1L, "F" = 2L)[sex],
      "age" = age,
      "bmi" = bmi,
      "rnaseq_path" = file.path(run_directory, "...")
    )]
    
    relatedness <- setDT(read_excel(exclusion, 2))[gsub("_.*", "", IID1) != gsub("_.*", "", IID2)]
    relatedness[j = paste0("vcf_id", 1:2) := list(paste0(FID1, "_", IID1), paste0(FID2, "_", IID2))]
    
    sample_sheet_qc <- sample_sheet_qc[ # not related
      !vcf_id %in% unique(unlist(relatedness[j = c("vcf_id1", "vcf_id2")], use.names = FALSE))
    ][ # European descent
      super_pop_closest %in% "EUR"
    ]
    
    fwrite(
      x = transpose(sample_sheet_qc[j = c("vcf_id", "sex", "age", "bmi", "t2d", "PC01", "PC02")], keep.names = "row"),
      file = file.path(output_covariates, "covariates.txt.gz"),
      quote = FALSE, col.names = FALSE, sep = "\t"
    )
    
    
### Omni2.5 data ===================================================================================
    vcfs <- list.files(
      path = file.path(vep_directory, "vcfs_qc"),
      pattern = "[^X].vcf.gz$",
      full.names = TRUE
    )
    names(vcfs) <- gsub("_qc.vcf.gz$", "", basename(vcfs))
    
    
### RNAseq data ====================================================================================
    rsem_files <- setNames(sample_sheet_qc[["rsem_file"]], sample_sheet_qc[["vcf_id"]])
    txi_counts <- tximport(
      files = rsem_files, 
      type = "rsem", 
      txIn = rna_level == "isoforms", 
      txOut = rna_level == "isoforms", 
      countsFromAbundance = "no"
    )
    txi_counts$length[txi_counts$length == 0] <- 1
    txi_counts[["counts"]] <- round(txi_counts[["counts"]], digits = 0)
    class(txi_counts[["counts"]]) <- "integer"
    gene_matrix <- txi_counts[["counts"]][
      rowVars(txi_counts[["counts"]]) != 0 & 
        rowMeans(txi_counts[["counts"]]) > 1 & 
        rowMedians(txi_counts[["counts"]]) > 0, 
    ]
    gene_matrix <- as.data.table(vst(gene_matrix))[j = ensembl_id := rownames(gene_matrix)]
    
    if (file.exists(file.path(output_rnaseq, "rnaseq_qc_map.csv.gz"))) {
      rnaseq_qc_map <- fread(file.path(output_rnaseq, "rnaseq_qc_map.csv.gz"))
    } else {
      rnaseq_qc_map <- setDT(getBM(
        attributes = c(glue("ensembl_{rna_level_name}_id"), "chromosome_name", "start_position", "end_position", "external_gene_name"),
        filters = glue("ensembl_{rna_level_name}_id"),
        values = list(gene_matrix[["ensembl_id"]]),
        mart = mart
      ))[
        j = lapply(.SD, function(x) {
          out <- paste(setdiff(unique(x), ""), collapse = ";")
          fifelse(out == "", NA_character_, out)
        }),
        by = c(glue("ensembl_{rna_level_name}_id"))
      ][j = ensembl_version := ensembl_version]
      fwrite(rnaseq_qc_map, file = file.path(output_rnaseq, "rnaseq_qc_map.csv.gz"))
    }
    
    rnaseq_qc <- merge(
      x = gene_matrix, 
      y = rnaseq_qc_map[j = -"external_gene_name"], 
      by.x = "ensembl_id",
      by.y = glue("ensembl_{rna_level_name}_id")
    )[
      chromosome_name %in% 1:22
    ][
      j = c("grp", "#Chr", "start", "end") := list(
        as.numeric(chromosome_name),
        as.numeric(chromosome_name),
        start_position,
        end_position
      )
    ][
      j = .SD,
      .SDcols = c("grp", "#Chr", "start", "end", "ensembl_id", setdiff(colnames(gene_matrix), "ensembl_id"))
    ][order(grp, start)]
    
    rnaseq_qc[
      j = (function(x, y) {
        fwrite(x, file = file.path(output_rnaseq, sprintf("chr%02d.bed", unique(y))), quote = FALSE, sep = "\t")
        system(paste("bgzip -f", file.path(output_rnaseq, sprintf("chr%02d.bed", unique(y)))))
        system(paste("tabix -p bed -f", file.path(output_rnaseq, sprintf("chr%02d.bed.gz", unique(y)))))
        return(TRUE)
      })(.SD, `#Chr`), 
      by = grp
    ]
    
    
### eQTL analysis (fastQTL) ========================================================================
    file_con <- gzfile(file.path(tempdir(), "nominal_header.txt.gz"), "w")
    cat(
      c("ensembl_id", "rs_id", "distance_bp", "pvalue", "slope\n"),
      sep = " ",
      file = file_con
    )
    close(file_con)
    
    file_con <- gzfile(file.path(tempdir(), "permutation_header.txt.gz"))
    cat(
      c(
        "ensembl_id", "variants_cis", "mle_shape1_beta", "mle_shape2_beta",
        "dummy", "best_rs_id", "distance_bp", "pvalue", "slope",
        "permutation_pvalue", "downstream_pvalue\n"
      ),
      sep = " ",
      file = file_con
    )
    close(file_con)
    
    for (ichr in sprintf("chr%02d", 1:22)) {
      local({
        for (ianalysis in c("nominal", "permutation")) {
          n_chunk <- 20
          mclapply(1:n_chunk, mc.cores = n_chunk, function(ichunk) {
            system(paste("fastQTL",
              "--silent",
              "--seed", seed,
              "--vcf", vcfs[ichr],
              "--bed", file.path(output_rnaseq, paste0(ichr, ".bed.gz")),
              "--cov", file.path(output_covariates, "covariates.txt.gz"),
              "--window", cis_window,
              ifelse(ianalysis == "nominal", "", "--permute 1000 10000"),
              "--chunk", ichunk, n_chunk,
              "--out", file.path(output_fastqtl, sprintf("%s_%s_%03d.txt.gz", ianalysis, ichr, ichunk))
            ))
          })
      
          system(paste(
            "zcat",
            file.path(tempdir(), sprintf("%s_header.txt.gz", ianalysis)),
            file.path(output_fastqtl, sprintf("%s_%s_*.txt.gz", ianalysis, ichr)),
            "| bgzip -c >", file.path(output_fastqtl, sprintf("%s_%s_%s.txt.gz", project_name, ianalysis, ichr))
          ))

          unlink(list.files(
            path = output_fastqtl,
            pattern = sprintf("%s_%s_[0-9]*.txt.gz", ianalysis, ichr),
            full.names = TRUE
          ))

          fwrite(
            x = merge(
              x = fread(file.path(output_fastqtl, sprintf("%s_%s_%s.txt.gz", project_name, ianalysis, ichr))),
              y = rnaseq_qc_map,
              by.x = "ensembl_id",
              by.y = glue("ensembl_{rna_level_name}_id"),
              all.x = TRUE
            ),
            file = file.path(output_fastqtl_annotated, sprintf("%s_%s_%s.txt.gz", project_name, ianalysis, ichr))
          )
        }
      })
    }


### Combine Results ================================================================================
    for (ianalysis in c("nominal", "permutation")) {
      local({
        fwrite(
          x = rbindlist(mclapply(
            X = sprintf("chr%02d", 1:22),
            mc.cores = 11, 
            mc.preschedule = FALSE,
            FUN = function(ichr) {
              fread(file.path(output_fastqtl_annotated, sprintf("%s_%s_%s.txt.gz", project_name, ianalysis, ichr)))
            }
          )),
          file = file.path(output_fastqtl_combined, sprintf("%s_%s.txt.gz", project_name, ianalysis))
        )
      })
    }


### Clean Directory ================================================================================
    unlink(c(
      gzfile(file.path(tempdir(), "nominal_header.txt.gz")),
      gzfile(file.path(tempdir(), "permutation_header.txt.gz"))
    ))
    unlink(setdiff(
      list.files(output_directory, full.names = TRUE),
      list.files(output_directory, pattern = "fastqtl_combined", full.names = TRUE)
    ), recursive = TRUE, force = TRUE)
  })
}


### Archive ========================================================================================
# if (!Sys.getenv("USER") %in% c("root", "") && file.exists("~/.fex/id")) {
#   local({
#     owd <- getwd()
#     setwd(normalizePath(output_directory))
#     archive_name <- file.path(
#       normalizePath(output_directory),
#       paste0(
#         format(Sys.Date(), format = "%Y%m%d"), "_",
#         project_name, "_", 
#         gsub("[0-9]+\\-", "", basename(output_directory)), ".zip"
#       )
#     )
#     zip(archive_name, files = list.files())
#     fex_out <- system(paste("fexsend", archive_name, "."), intern = TRUE)
#     unlink(archive_name)
#     setwd(owd)
#   })
# }


### Set chmod ======================================================================================
Sys.chmod(
  list.files(output_directory, full.names = TRUE),
  mode = "0775", use_umask = FALSE
)
Sys.chmod(
  list.files(output_directory, full.names = TRUE, recursive = TRUE, all.files = TRUE),
  mode = "0775", use_umask = FALSE
)
invisible(system(paste("chgrp -R staff", output_directory), intern = TRUE))


### Complete =======================================================================================
message("Success!", appendLF = TRUE)
message(timestamp(quiet = TRUE))
